File: Stack/BOJ_2493.py
N = int(input())
topList = list(map(int,input().split()))
stack = []
answer = []

for i in range(N):
  while stack:
    if topList[stack[-1]] >= topList[i]:
      break
    else:
      stack.pop()

  if stack:
    answer.append(stack[-1] + 1)
  else:
    answer.append(0)

  stack.append(i)

print(' '.join(map(str,answer)))

File: Stack/BOJ_2493.js
let fs = require('fs')
let input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const N = Number(input[0]);
const topList = input[1].split(' ')
const stack = []
const answer = []

for (let i = 0; i < N; i++) {
  while (stack.length) {
    if (Number(topList[stack[stack.length-1]]) >= Number(topList[i])) {
      break
    } else {
      stack.pop()
    }
  }

  if (stack.length) {
    answer.push(stack[stack.length-1] + 1)
  } else {
    answer.push(0)
  }

  stack.push(i)
}

console.log(answer.join(' '))